work out why CS8631 and CS8620 started

with an update to dotnet stable we started getting the following build warnings `CS8631` `CS8620`, which resulted in failing building in release mode.

```
C:\projects\shouldly\src\Shouldly\Internals\Is.cs(29,44): error CS8620: Argument of type 'IEqualityComparer<T>' cannot be used for parameter 'comparer' of type 'IEqualityComparer<T?>' in 'bool Is.Equal<T?>(T? expected, T? actual, IEqualityComparer<T?> comparer)' due to differences in the nullability of reference types. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\Internals\Is.cs(282,20): error CS8631: The type 'T?' cannot be used as type parameter 'T' in the generic type or method 'Is.Compare<T>(T, T)'. Nullability of type argument 'T?' doesn't match constraint type 'System.IComparable<T?>?'. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\Internals\Is.cs(287,46): error CS8620: Argument of type 'IComparer<T>' cannot be used for parameter 'comparer' of type 'IComparer<T?>' in 'decimal Is.Compare<T?>(T? actual, T? expected, IComparer<T?> comparer)' due to differences in the nullability of reference types. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\Internals\Is.cs(292,20): error CS8631: The type 'T?' cannot be used as type parameter 'T' in the generic type or method 'Is.Compare<T>(T, T)'. Nullability of type argument 'T?' doesn't match constraint type 'System.IComparable<T?>?'. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\Internals\Is.cs(297,46): error CS8620: Argument of type 'IComparer<T>' cannot be used for parameter 'comparer' of type 'IComparer<T?>' in 'decimal Is.Compare<T?>(T? actual, T? expected, IComparer<T?> comparer)' due to differences in the nullability of reference types. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\Internals\Is.cs(302,20): error CS8631: The type 'T?' cannot be used as type parameter 'T' in the generic type or method 'Is.Compare<T>(T, T)'. Nullability of type argument 'T?' doesn't match constraint type 'System.IComparable<T?>?'. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\Internals\Is.cs(307,46): error CS8620: Argument of type 'IComparer<T>' cannot be used for parameter 'comparer' of type 'IComparer<T?>' in 'decimal Is.Compare<T?>(T? actual, T? expected, IComparer<T?> comparer)' due to differences in the nullability of reference types. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\Internals\Is.cs(312,20): error CS8631: The type 'T?' cannot be used as type parameter 'T' in the generic type or method 'Is.Compare<T>(T, T)'. Nullability of type argument 'T?' doesn't match constraint type 'System.IComparable<T?>?'. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\Internals\Is.cs(317,46): error CS8620: Argument of type 'IComparer<T>' cannot be used for parameter 'comparer' of type 'IComparer<T?>' in 'decimal Is.Compare<T?>(T? actual, T? expected, IComparer<T?> comparer)' due to differences in the nullability of reference types. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\ShouldlyExtensionMethods\ShouldBeGtLtTestExtensions.cs(11,75): error CS8620: Argument of type 'IComparer<T>' cannot be used for parameter 'comparer' of type 'IComparer<T?>' in 'bool Is.GreaterThan<T?>(T? actual, T? expected, IComparer<T?> comparer)' due to differences in the nullability of reference types. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\ShouldlyExtensionMethods\ShouldBeGtLtTestExtensions.cs(16,42): error CS8631: The type 'T?' cannot be used as type parameter 'T' in the generic type or method 'Is.GreaterThan<T>(T, T)'. Nullability of type argument 'T?' doesn't match constraint type 'System.IComparable<T?>?'. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\ShouldlyExtensionMethods\ShouldBeGtLtTestExtensions.cs(21,72): error CS8620: Argument of type 'IComparer<T>' cannot be used for parameter 'comparer' of type 'IComparer<T?>' in 'bool Is.LessThan<T?>(T? actual, T? expected, IComparer<T?> comparer)' due to differences in the nullability of reference types. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\ShouldlyExtensionMethods\ShouldBeGtLtTestExtensions.cs(26,42): error CS8631: The type 'T?' cannot be used as type parameter 'T' in the generic type or method 'Is.LessThan<T>(T, T)'. Nullability of type argument 'T?' doesn't match constraint type 'System.IComparable<T?>?'. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\ShouldlyExtensionMethods\ShouldBeGtLtTestExtensions.cs(31,84): error CS8620: Argument of type 'IComparer<T>' cannot be used for parameter 'comparer' of type 'IComparer<T?>' in 'bool Is.GreaterThanOrEqualTo<T?>(T? actual, T? expected, IComparer<T?> comparer)' due to differences in the nullability of reference types. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\ShouldlyExtensionMethods\ShouldBeGtLtTestExtensions.cs(36,42): error CS8631: The type 'T?' cannot be used as type parameter 'T' in the generic type or method 'Is.GreaterThanOrEqualTo<T>(T, T)'. Nullability of type argument 'T?' doesn't match constraint type 'System.IComparable<T?>?'. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\ShouldlyExtensionMethods\ShouldBeGtLtTestExtensions.cs(41,81): error CS8620: Argument of type 'IComparer<T>' cannot be used for parameter 'comparer' of type 'IComparer<T?>' in 'bool Is.LessThanOrEqualTo<T?>(T? actual, T? expected, IComparer<T?> comparer)' due to differences in the nullability of reference types. [C:\projects\shouldly\src\Shouldly\Shouldly.csproj]
C:\projects\shouldly\src\Shouldly\ShouldlyExtensionMethods\ShouldBeGtLtTestExtensions.cs(46,42): error CS8631: The type 'T?' cannot be used as type parameter 'T' in the generic type or method 'Is.LessThanOrEqualTo<T>(T, T)'. Nullability of type argument 'T?' doesn't 
```

--- a/src/Directory.Build.props
+++ b/src/Directory.Build.props
@@ -4,7 +4,6 @@
     <LangVersion>latest</LangVersion>
     <Nullable>enable</Nullable>
     <Version>4.0.1</Version>
-    <NoWarn>CS8631;CS8620</NoWarn>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
   </PropertyGroup>
 

--- a/src/Shouldly.Tests/ConventionTests/ApprovePublicApi.ShouldlyApi.approved.cs
+++ b/src/Shouldly.Tests/ConventionTests/ApprovePublicApi.ShouldlyApi.approved.cs
@@ -531,12 +531,12 @@ namespace Shouldly
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBe<T>(this T actual, T expected, System.Func<string?>? customMessage) { }
-        public static void ShouldBe<T>([System.Diagnostics.CodeAnalysis.AllowNull] [System.Diagnostics.CodeAnalysis.NotNullIfNotNull("expected")] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] [System.Diagnostics.CodeAnalysis.NotNullIfNotNull("actual")] T expected, string? customMessage = null) { }
+        public static void ShouldBe<T>([System.Diagnostics.CodeAnalysis.NotNullIfNotNull("expected")] this T? actual, [System.Diagnostics.CodeAnalysis.NotNullIfNotNull("actual")] T? expected, string? customMessage = null) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBe<T>(this System.Collections.Generic.IEnumerable<T>? actual, System.Collections.Generic.IEnumerable<T>? expected, bool ignoreOrder, System.Func<string?>? customMessage) { }
         public static void ShouldBe<T>([System.Diagnostics.CodeAnalysis.NotNullIfNotNull("expected")] this System.Collections.Generic.IEnumerable<T>? actual, [System.Diagnostics.CodeAnalysis.NotNullIfNotNull("actual")] System.Collections.Generic.IEnumerable<T>? expected, bool ignoreOrder, string? customMessage) { }
-        public static void ShouldBe<T>([System.Diagnostics.CodeAnalysis.AllowNull] [System.Diagnostics.CodeAnalysis.NotNullIfNotNull("expected")] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] [System.Diagnostics.CodeAnalysis.NotNullIfNotNull("actual")] T expected, System.Collections.Generic.IEqualityComparer<T> comparer, string? customMessage = null) { }
+        public static void ShouldBe<T>([System.Diagnostics.CodeAnalysis.NotNullIfNotNull("expected")] this T? actual, [System.Diagnostics.CodeAnalysis.NotNullIfNotNull("actual")] T? expected, System.Collections.Generic.IEqualityComparer<T> comparer, string? customMessage = null) { }
         public static void ShouldBe<T>([System.Diagnostics.CodeAnalysis.NotNullIfNotNull("expected")] this System.Collections.Generic.IEnumerable<T>? actual, [System.Diagnostics.CodeAnalysis.NotNullIfNotNull("actual")] System.Collections.Generic.IEnumerable<T>? expected, System.Collections.Generic.IEqualityComparer<T> comparer, bool ignoreOrder = false, string? customMessage = null) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
@@ -550,47 +550,47 @@ namespace Shouldly
             "stomMessage.", true)]
         public static void ShouldBeGreaterThan<T>(this T actual, T expected, System.Func<string?>? customMessage)
             where T : System.IComparable<T>? { }
-        public static void ShouldBeGreaterThan<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T expected, string? customMessage = null)
+        public static void ShouldBeGreaterThan<T>(this T? actual, T? expected, string? customMessage = null)
             where T : System.IComparable<T>? { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBeGreaterThan<T>(this T actual, T expected, System.Collections.Generic.IComparer<T> comparer, System.Func<string?>? customMessage) { }
-        public static void ShouldBeGreaterThan<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T expected, System.Collections.Generic.IComparer<T> comparer, string? customMessage = null) { }
+        public static void ShouldBeGreaterThan<T>(this T? actual, T? expected, System.Collections.Generic.IComparer<T> comparer, string? customMessage = null) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBeGreaterThanOrEqualTo<T>(this T actual, T expected, System.Func<string?>? customMessage)
             where T : System.IComparable<T>? { }
-        public static void ShouldBeGreaterThanOrEqualTo<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T expected, string? customMessage = null)
+        public static void ShouldBeGreaterThanOrEqualTo<T>(this T? actual, T? expected, string? customMessage = null)
             where T : System.IComparable<T>? { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBeGreaterThanOrEqualTo<T>(this T actual, T expected, System.Collections.Generic.IComparer<T> comparer, System.Func<string?>? customMessage) { }
-        public static void ShouldBeGreaterThanOrEqualTo<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T expected, System.Collections.Generic.IComparer<T> comparer, string? customMessage = null) { }
+        public static void ShouldBeGreaterThanOrEqualTo<T>(this T? actual, T? expected, System.Collections.Generic.IComparer<T> comparer, string? customMessage = null) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBeInRange<T>(this T actual, T from, T to, System.Func<string?>? customMessage) { }
-        public static void ShouldBeInRange<T>([System.Diagnostics.CodeAnalysis.DisallowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T from, [System.Diagnostics.CodeAnalysis.AllowNull] T to, string? customMessage = null)
+        public static void ShouldBeInRange<T>([System.Diagnostics.CodeAnalysis.DisallowNull] this T actual, T? from, T? to, string? customMessage = null)
             where T : System.IComparable<T> { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBeLessThan<T>(this T actual, T expected, System.Func<string?>? customMessage)
             where T : System.IComparable<T>? { }
-        public static void ShouldBeLessThan<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T expected, string? customMessage = null)
+        public static void ShouldBeLessThan<T>(this T? actual, T? expected, string? customMessage = null)
             where T : System.IComparable<T>? { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBeLessThan<T>(this T actual, T expected, System.Collections.Generic.IComparer<T> comparer, System.Func<string?>? customMessage) { }
-        public static void ShouldBeLessThan<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T expected, System.Collections.Generic.IComparer<T> comparer, string? customMessage = null) { }
+        public static void ShouldBeLessThan<T>(this T? actual, T? expected, System.Collections.Generic.IComparer<T> comparer, string? customMessage = null) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBeLessThanOrEqualTo<T>(this T actual, T expected, System.Func<string?>? customMessage)
             where T : System.IComparable<T>? { }
-        public static void ShouldBeLessThanOrEqualTo<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T expected, string? customMessage = null)
+        public static void ShouldBeLessThanOrEqualTo<T>(this T? actual, T? expected, string? customMessage = null)
             where T : System.IComparable<T>? { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBeLessThanOrEqualTo<T>(this T actual, T expected, System.Collections.Generic.IComparer<T> comparer, System.Func<string?>? customMessage) { }
-        public static void ShouldBeLessThanOrEqualTo<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T expected, System.Collections.Generic.IComparer<T> comparer, string? customMessage = null) { }
+        public static void ShouldBeLessThanOrEqualTo<T>(this T? actual, T? expected, System.Collections.Generic.IComparer<T> comparer, string? customMessage = null) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBeNegative(this decimal actual, System.Func<string?>? customMessage) { }
@@ -623,12 +623,12 @@ namespace Shouldly
             "stomMessage.", true)]
         public static T ShouldBeOfType<T>(this object? actual, System.Func<string?>? customMessage) { }
         public static T ShouldBeOfType<T>([System.Diagnostics.CodeAnalysis.NotNull] this object? actual, string? customMessage = null) { }
-        public static void ShouldBeOneOf<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, params T[] expected) { }
+        public static void ShouldBeOneOf<T>(this T? actual, params T[] expected) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBeOneOf<T>(this T actual, T[] expected, System.Func<string?>? customMessage) { }
-        public static void ShouldBeOneOf<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, T[] expected, string? customMessage) { }
-        public static void ShouldBeOneOf<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, T[] expected, System.Collections.Generic.IEqualityComparer<T> comparer, string? customMessage = null) { }
+        public static void ShouldBeOneOf<T>(this T? actual, T[] expected, string? customMessage) { }
+        public static void ShouldBeOneOf<T>(this T? actual, T[] expected, System.Collections.Generic.IEqualityComparer<T> comparer, string? customMessage = null) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBePositive(this decimal actual, System.Func<string?>? customMessage) { }
@@ -672,8 +672,8 @@ namespace Shouldly
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldNotBe<T>(this T actual, T expected, System.Func<string?>? customMessage) { }
-        public static void ShouldNotBe<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T expected, string? customMessage = null) { }
-        public static void ShouldNotBe<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T expected, System.Collections.Generic.IEqualityComparer<T> comparer, string? customMessage = null) { }
+        public static void ShouldNotBe<T>(this T? actual, T? expected, string? customMessage = null) { }
+        public static void ShouldNotBe<T>(this T? actual, T? expected, System.Collections.Generic.IEqualityComparer<T> comparer, string? customMessage = null) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldNotBeAssignableTo(this object? actual, System.Type expected, System.Func<string?>? customMessage) { }
@@ -686,7 +686,7 @@ namespace Shouldly
             "stomMessage.", true)]
         public static void ShouldNotBeInRange<T>(this T actual, T from, T to, System.Func<string?>? customMessage)
             where T : System.IComparable<T> { }
-        public static void ShouldNotBeInRange<T>([System.Diagnostics.CodeAnalysis.DisallowNull] this T actual, [System.Diagnostics.CodeAnalysis.AllowNull] T from, [System.Diagnostics.CodeAnalysis.AllowNull] T to, string? customMessage = null)
+        public static void ShouldNotBeInRange<T>([System.Diagnostics.CodeAnalysis.DisallowNull] this T actual, T? from, T? to, string? customMessage = null)
             where T : System.IComparable<T> { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
@@ -696,12 +696,12 @@ namespace Shouldly
             "stomMessage.", true)]
         public static void ShouldNotBeOfType<T>(this object? actual, System.Func<string?>? customMessage) { }
         public static void ShouldNotBeOfType<T>(this object? actual, string? customMessage = null) { }
-        public static void ShouldNotBeOneOf<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, params T[] expected) { }
+        public static void ShouldNotBeOneOf<T>(this T? actual, params T[] expected) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldNotBeOneOf<T>(this T actual, T[] expected, System.Func<string?>? customMessage) { }
-        public static void ShouldNotBeOneOf<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, T[] expected, string? customMessage) { }
-        public static void ShouldNotBeOneOf<T>([System.Diagnostics.CodeAnalysis.AllowNull] this T actual, T[] expected, System.Collections.Generic.IEqualityComparer<T> comparer, string? customMessage = null) { }
+        public static void ShouldNotBeOneOf<T>(this T? actual, T[] expected, string? customMessage) { }
+        public static void ShouldNotBeOneOf<T>(this T? actual, T[] expected, System.Collections.Generic.IEqualityComparer<T> comparer, string? customMessage = null) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldNotBeSameAs(this object? actual, object? expected, System.Func<string?>? customMessage) { }

--- a/src/Shouldly.Tests/ConventionTests/ApprovePublicApi.ShouldlyApi.approved.cs
+++ b/src/Shouldly.Tests/ConventionTests/ApprovePublicApi.ShouldlyApi.approved.cs
@@ -545,7 +545,7 @@ namespace Shouldly
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static T ShouldBeAssignableTo<T>(this object? actual, System.Func<string?>? customMessage) { }
-        public static T ShouldBeAssignableTo<T>(this object? actual, string? customMessage = null) { }
+        public static T? ShouldBeAssignableTo<T>(this object? actual, string? customMessage = null) { }
         [System.Obsolete("Func based customMessage overloads have been removed. Pass in a string for the cu" +
             "stomMessage.", true)]
         public static void ShouldBeGreaterThan<T>(this T actual, T expected, System.Func<string?>? customMessage)

--- a/src/Shouldly.Tests/TestHelpers/ComparableClassComparer.cs
+++ b/src/Shouldly.Tests/TestHelpers/ComparableClassComparer.cs
@@ -14,7 +14,7 @@ namespace Shouldly.Tests.TestHelpers
 
         public int GetHashCode(ComparableClass obj)
         {
-            return EqualityComparer<string?>.Default.GetHashCode(obj.Property!);
+            return EqualityComparer<string?>.Default.GetHashCode(obj.Property);
         }
     }
 }

--- a/src/Shouldly.Tests/TestHelpers/CustomComparerClass.cs
+++ b/src/Shouldly.Tests/TestHelpers/CustomComparerClass.cs
@@ -1,11 +1,10 @@
 ﻿using System.Collections.Generic;
-using System.Diagnostics.CodeAnalysis;
 
 namespace Shouldly.Tests.TestHelpers
 {
     internal class CustomComparer<T> : IComparer<T>
     {
-        public int Compare([AllowNull] T x, [AllowNull] T y)
+        public int Compare(T? x, T? y)
         {
             Custom x1 = (Custom)(object)x!;
             Custom x2 = (Custom)(object)y!;

--- a/src/Shouldly/Configuration/FindMethodUsingAttribute.cs
+++ b/src/Shouldly/Configuration/FindMethodUsingAttribute.cs
@@ -9,7 +9,7 @@ namespace Shouldly.Configuration
     {
         public TestMethodInfo GetTestMethodInfo(StackTrace stackTrace, int startAt = 0)
         {
-            for (var i = startAt; stackTrace.GetFrame(i) is { } frame; i++)
+            foreach (var frame in stackTrace.GetFrames().Skip(startAt))
             {
                 var method = frame.GetMethod();
                 var originalMethod = GetOriginalMethodInfoForStateMachineMethod(method);

--- a/src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs
+++ b/src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs
@@ -1,5 +1,7 @@
+using Shouldly.Internals;
 using System;
 using System.Diagnostics;
+using System.Linq;
 using System.Reflection;
 using System.Runtime.CompilerServices;
 using System.Text.RegularExpressions;
@@ -19,7 +21,7 @@ namespace Shouldly.Configuration
 
         public TestMethodInfo GetTestMethodInfo(StackTrace stackTrace, int startAt = 0)
         {
-            for (var i = startAt; stackTrace.GetFrame(i) is { } frame; i++)
+            foreach (var (i, frame) in stackTrace.GetFrames().AsIndexed().Skip(startAt))
             {
                 if (frame.GetMethod() is { } method && !method.IsShouldlyMethod() && !IsCompilerGenerated(method))
                 {

--- a/src/Shouldly/DifferenceHighlighting/EnumerableDifferenceHighlighter.cs
+++ b/src/Shouldly/DifferenceHighlighting/EnumerableDifferenceHighlighter.cs
@@ -34,8 +34,8 @@ namespace Shouldly.DifferenceHighlighting
             var expected = context.Expected as IEnumerable;
             if (CanProcess(context))
             {
-                var actualList = actual.Cast<object>();
-                var expectedList = expected.Cast<object>();
+                var actualList = actual!.Cast<object>();
+                var expectedList = expected!.Cast<object>();
 
                 var highestCount = actualList.Count() > expectedList.Count() ? actualList.Count() : expectedList.Count();
 

--- a/src/Shouldly/Internals/EnumerableExtensions.cs
+++ b/src/Shouldly/Internals/EnumerableExtensions.cs
@@ -1,0 +1,18 @@
+﻿using System.Collections.Generic;
+
+namespace Shouldly.Internals
+{
+    internal static class EnumerableExtensions
+    {
+        public static IEnumerable<(int Index, T Value)> AsIndexed<T>(this IEnumerable<T> source)
+        {
+            var index = 0;
+
+            foreach (var value in source)
+            {
+                yield return (index, value);
+                index++;
+            }
+        }
+    }
+}

--- a/src/Shouldly/Internals/EqualityComparer.cs
+++ b/src/Shouldly/Internals/EqualityComparer.cs
@@ -1,7 +1,6 @@
 using System;
 using System.Collections;
 using System.Collections.Generic;
-using System.Diagnostics.CodeAnalysis;
 using NUnit.Framework.Constraints;
 
 namespace Shouldly
@@ -22,7 +21,7 @@ namespace Shouldly
             _innerComparerFactory = () => innerComparer ?? DefaultInnerComparer;
         }
 
-        public bool Equals([AllowNull] T x, [AllowNull] T y)
+        public bool Equals(T? x, T? y)
         {
             if (ReferenceEquals(x, y))
                 return true;

--- a/src/Shouldly/Internals/Is.cs
+++ b/src/Shouldly/Internals/Is.cs
@@ -9,7 +9,7 @@ namespace Shouldly
 {
     internal static class Is
     {
-        public static bool InRange<T>([DisallowNull] T comparable, [AllowNull] T @from, [AllowNull] T to) where T : IComparable<T>
+        public static bool InRange<T>([DisallowNull] T comparable, T? @from, T? to) where T : IComparable<T>
         {
             return comparable.CompareTo(from) >= 0 && comparable.CompareTo(to) <= 0;
         }
@@ -24,12 +24,12 @@ namespace Shouldly
             return ReferenceEquals(actual, expected);
         }
 
-        public static bool Equal<T>([AllowNull] T expected, [AllowNull] T actual)
+        public static bool Equal<T>(T? expected, T? actual)
         {
             return Equal(expected, actual, GetEqualityComparer<T>());
         }
 
-        public static bool Equal<T>([AllowNull] T expected, [AllowNull] T actual, IEqualityComparer<T> comparer)
+        public static bool Equal<T>(T? expected, T? actual, IEqualityComparer<T> comparer)
         {
             return comparer.Equals(actual, expected);
         }
@@ -277,52 +277,52 @@ namespace Shouldly
             }
         }
 
-        public static bool GreaterThanOrEqualTo<T>([AllowNull] T comparable, [AllowNull] T expected) where T : IComparable<T>?
+        public static bool GreaterThanOrEqualTo<T>(T? comparable, T? expected) where T : IComparable<T>?
         {
             return Compare(comparable, expected) >= 0;
         }
 
-        public static bool GreaterThanOrEqualTo<T>([AllowNull] T actual, [AllowNull] T expected, IComparer<T> comparer)
+        public static bool GreaterThanOrEqualTo<T>(T? actual, T? expected, IComparer<T> comparer)
         {
             return Compare(actual, expected, comparer) >= 0;
         }
 
-        public static bool LessThanOrEqualTo<T>([AllowNull] T comparable, [AllowNull] T expected) where T : IComparable<T>?
+        public static bool LessThanOrEqualTo<T>(T? comparable, T? expected) where T : IComparable<T>?
         {
             return Compare(comparable, expected) <= 0;
         }
 
-        public static bool LessThanOrEqualTo<T>([AllowNull] T actual, [AllowNull] T expected, IComparer<T> comparer)
+        public static bool LessThanOrEqualTo<T>(T? actual, T? expected, IComparer<T> comparer)
         {
             return Compare(actual, expected, comparer) <= 0;
         }
 
-        public static bool GreaterThan<T>([AllowNull] T comparable, [AllowNull] T expected) where T : IComparable<T>?
+        public static bool GreaterThan<T>(T? comparable, T? expected) where T : IComparable<T>?
         {
             return Compare(comparable, expected) > 0;
         }
 
-        public static bool GreaterThan<T>([AllowNull] T actual, [AllowNull] T expected, IComparer<T> comparer)
+        public static bool GreaterThan<T>(T? actual, T? expected, IComparer<T> comparer)
         {
             return Compare(actual, expected, comparer) > 0;
         }
 
-        public static bool LessThan<T>([AllowNull] T comparable, [AllowNull] T expected) where T : IComparable<T>?
+        public static bool LessThan<T>(T? comparable, T? expected) where T : IComparable<T>?
         {
             return Compare(comparable, expected) < 0;
         }
 
-        public static bool LessThan<T>([AllowNull] T actual, [AllowNull] T expected, IComparer<T> comparer)
+        public static bool LessThan<T>(T? actual, T? expected, IComparer<T> comparer)
         {
             return Compare(actual, expected, comparer) < 0;
         }
 
-        static decimal Compare<T>([AllowNull] T actual, [AllowNull] T expected, IComparer<T> comparer)
+        static decimal Compare<T>(T? actual, T? expected, IComparer<T> comparer)
         {
             return comparer.Compare(actual, expected);
         }
 
-        static decimal Compare<T>([AllowNull] T comparable, [AllowNull] T expected) where T : IComparable<T>?
+        static decimal Compare<T>(T? comparable, T? expected) where T : IComparable<T>?
         {
             if (!typeof(T).IsValueType())
             {

--- a/src/Shouldly/Internals/Is.cs
+++ b/src/Shouldly/Internals/Is.cs
@@ -97,7 +97,7 @@ namespace Shouldly
             foreach (var actualElement in actual)
             {
                 var match = expectedList.FirstOrDefault(x => Equal(x, actualElement, comparer));
-                if (!expectedList.Remove(match))
+                if (!expectedList.Remove(match!)) // List<T>.Remove works fine when null is passed.
                     return false;
             }
 

--- a/src/Shouldly/Internals/ObjectEqualityComparer.cs
+++ b/src/Shouldly/Internals/ObjectEqualityComparer.cs
@@ -1,12 +1,11 @@
 ﻿using System;
-using System.Diagnostics.CodeAnalysis;
 
 namespace Shouldly
 {
     [Serializable]
     class ObjectEqualityComparer<T> : System.Collections.Generic.EqualityComparer<T>
     {
-        public override bool Equals([AllowNull] T x, [AllowNull] T y)
+        public override bool Equals(T? x, T? y)
         {
             if (x != null)
             {
@@ -15,7 +14,7 @@ namespace Shouldly
             return y == null;
         }
 
-        public override int GetHashCode([AllowNull] T obj)
+        public override int GetHashCode(T? obj)
         {
             if (obj == null)
             {

--- a/src/Shouldly/Internals/SourceCodeTextGetter.cs
+++ b/src/Shouldly/Internals/SourceCodeTextGetter.cs
@@ -60,7 +60,7 @@ namespace Shouldly.Internals
 
             if (_determinedOriginatingFrame)
             {
-                var codeLines = string.Join("\n", File.ReadAllLines(FileName).Skip(LineNumber).ToArray());
+                var codeLines = string.Join("\n", File.ReadAllLines(FileName!).Skip(LineNumber).ToArray());
 
                 var indexOf = codeLines.IndexOf(_shouldMethod!);
                 if (indexOf > 0)

--- a/src/Shouldly/Internals/StackTraceHelpers.cs
+++ b/src/Shouldly/Internals/StackTraceHelpers.cs
@@ -24,9 +24,9 @@ namespace Shouldly.Internals
         {
             var shouldlyAssembly = Assembly.GetExecutingAssembly();
 
-            for (var startIndex = 0; startIndex < stackTrace.FrameCount; startIndex++)
+            var frames = stackTrace.GetFrames();
+            foreach (var (startIndex, frame) in frames.AsIndexed())
             {
-                var frame = stackTrace.GetFrame(startIndex)!;
                 if (frame.GetMethod()?.DeclaringType?.Assembly != shouldlyAssembly)
                 {
                     if (startIndex == 0)
@@ -35,12 +35,12 @@ namespace Shouldly.Internals
                     }
                     else
                     {
-                        var lines = new string[stackTrace.FrameCount - startIndex];
+                        var lines = new string[frames.Length - startIndex];
                         var neededCapacity = builder.Length;
 
                         for (var i = 0; i < lines.Length; i++)
                         {
-                            var line = new StackTrace(stackTrace.GetFrame(i + startIndex)!).ToString();
+                            var line = new StackTrace(frames[i + startIndex]).ToString();
                             if (i == lines.Length - 1) line = line.TrimEnd();
                             lines[i] = line;
                             neededCapacity += line.Length;

--- a/src/Shouldly/Internals/StringHelpers.cs
+++ b/src/Shouldly/Internals/StringHelpers.cs
@@ -61,7 +61,7 @@ namespace Shouldly
             if (value is MemberExpression)
             {
                 var member = value.As<MemberExpression>();
-                var constant = member.Expression.As<ConstantExpression>();
+                var constant = member.Expression!.As<ConstantExpression>();
                 var info = member.Member.As<FieldInfo>();
                 return info.GetValue(constant.Value).ToStringAwesomely();
             }

--- a/src/Shouldly/Internals/TaskExtensions.cs
+++ b/src/Shouldly/Internals/TaskExtensions.cs
@@ -9,7 +9,7 @@ namespace Shouldly
     {
         private struct VoidTypeStruct { }
 
-        private static void MarshalTaskResults<TResult>(Task source, TaskCompletionSource<TResult> proxy)
+        private static void MarshalTaskResults<TResult>(Task source, TaskCompletionSource<TResult?> proxy)
         {
             switch (source.Status)
             {
@@ -21,7 +21,7 @@ namespace Shouldly
                     break;
                 case TaskStatus.RanToCompletion:
                     proxy.TrySetResult(!(source is Task<TResult> castedSource)
-                        ? default! // source is a Task
+                        ? default // source is a Task
                         : castedSource.Result); // source is a Task<TResult>
                     break;
             }

--- a/src/Shouldly/Internals/TaskExtensions.cs
+++ b/src/Shouldly/Internals/TaskExtensions.cs
@@ -14,7 +14,7 @@ namespace Shouldly
             switch (source.Status)
             {
                 case TaskStatus.Faulted:
-                    proxy.TrySetException(source.Exception!);
+                    proxy.TrySetException(source.Exception!.InnerExceptions);
                     break;
                 case TaskStatus.Canceled:
                     proxy.TrySetCanceled();

--- a/src/Shouldly/Shouldly.csproj
+++ b/src/Shouldly/Shouldly.csproj
@@ -29,6 +29,6 @@
     <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
     <None Include="..\..\assets\logo_128x128.png" Pack="true" PackagePath="assets" />
     <PackageReference Include="TunnelVisionLabs.ReferenceAssemblyAnnotator" Version="1.0.0-alpha.160" PrivateAssets="all" />
-    <PackageDownload Include="Microsoft.NETCore.App.Ref" Version="[3.1.0]" />
+    <PackageDownload Include="Microsoft.NETCore.App.Ref" Version="[5.0.0]" />
   </ItemGroup>
 </Project>

--- a/src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs
+++ b/src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs
@@ -11,8 +11,8 @@ namespace Shouldly
     {
         [ContractAnnotation("actual:null,expected:notnull => halt;actual:notnull,expected:null => halt")]
         public static void  ShouldBe<T>(
-            [AllowNull, NotNullIfNotNull("expected")] this T actual,
-            [AllowNull, NotNullIfNotNull("actual")] T expected,
+            [NotNullIfNotNull("expected")] this T? actual,
+            [NotNullIfNotNull("actual")] T? expected,
             string? customMessage = null)
         {
             if (ShouldlyConfiguration.CompareAsObjectTypes.Contains(typeof(T).FullName!) || typeof(T) == typeof(string))
@@ -22,8 +22,8 @@ namespace Shouldly
         }
 
         public static void ShouldBe<T>(
-            [AllowNull, NotNullIfNotNull("expected")] this T actual,
-            [AllowNull, NotNullIfNotNull("actual")] T expected,
+            [NotNullIfNotNull("expected")] this T? actual,
+            [NotNullIfNotNull("actual")] T? expected,
             IEqualityComparer<T> comparer,
             string? customMessage = null)
         {
@@ -31,13 +31,13 @@ namespace Shouldly
         }
 
         [ContractAnnotation("actual:null,expected:null => halt")]
-        public static void ShouldNotBe<T>([AllowNull] this T actual, [AllowNull] T expected, string? customMessage = null)
+        public static void ShouldNotBe<T>(this T? actual, T? expected, string? customMessage = null)
         {
             actual!.AssertAwesomely(v => !Is.Equal(v, expected), actual, expected, customMessage);
         }
 
         [ContractAnnotation("actual:null,expected:null => halt")]
-        public static void ShouldNotBe<T>([AllowNull] this T actual, [AllowNull] T expected, IEqualityComparer<T> comparer, string? customMessage = null)
+        public static void ShouldNotBe<T>(this T? actual, T? expected, IEqualityComparer<T> comparer, string? customMessage = null)
         {
             actual!.AssertAwesomely(v => !Is.Equal(v, expected, comparer), actual, expected, customMessage);
         }

--- a/src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs
+++ b/src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs
@@ -16,9 +16,9 @@ namespace Shouldly
             string? customMessage = null)
         {
             if (ShouldlyConfiguration.CompareAsObjectTypes.Contains(typeof(T).FullName!) || typeof(T) == typeof(string))
-                actual!.AssertAwesomely(v => Is.Equal(v, expected, new ObjectEqualityComparer<T>()), actual, expected, customMessage);
+                actual.AssertAwesomely(v => Is.Equal(v, expected, new ObjectEqualityComparer<T>()), actual, expected, customMessage);
             else
-                actual!.AssertAwesomely(v => Is.Equal(v, expected), actual, expected, customMessage);
+                actual.AssertAwesomely(v => Is.Equal(v, expected), actual, expected, customMessage);
         }
 
         public static void ShouldBe<T>(
@@ -27,19 +27,19 @@ namespace Shouldly
             IEqualityComparer<T> comparer,
             string? customMessage = null)
         {
-            actual!.AssertAwesomely(v => Is.Equal(v, expected, comparer), actual, expected, customMessage);
+            actual.AssertAwesomely(v => Is.Equal(v, expected, comparer), actual, expected, customMessage);
         }
 
         [ContractAnnotation("actual:null,expected:null => halt")]
         public static void ShouldNotBe<T>(this T? actual, T? expected, string? customMessage = null)
         {
-            actual!.AssertAwesomely(v => !Is.Equal(v, expected), actual, expected, customMessage);
+            actual.AssertAwesomely(v => !Is.Equal(v, expected), actual, expected, customMessage);
         }
 
         [ContractAnnotation("actual:null,expected:null => halt")]
         public static void ShouldNotBe<T>(this T? actual, T? expected, IEqualityComparer<T> comparer, string? customMessage = null)
         {
-            actual!.AssertAwesomely(v => !Is.Equal(v, expected, comparer), actual, expected, customMessage);
+            actual.AssertAwesomely(v => !Is.Equal(v, expected, comparer), actual, expected, customMessage);
         }
 
         public static void ShouldBe<T>(

--- a/src/Shouldly/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs
+++ b/src/Shouldly/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs
@@ -189,13 +189,13 @@ namespace Shouldly
 
         private static void ShouldBeInOrder<T>(IEnumerable<T> actual, SortDirection expectedSortDirection, Func<T, T, bool> isOutOfOrder, string? customMessage)
         {
-            var previousItem = default(T)!;
+            var previousItem = default(T);
             var currentIndex = -1;
 
             foreach (var currentItem in actual)
             {
                 if (++currentIndex > 0 // We only need to start comparing once we've passed the first item in the list
-                    && isOutOfOrder(previousItem, currentItem))
+                    && isOutOfOrder(previousItem!, currentItem))
                 {
                     throw new ShouldAssertException(
                         new ExpectedOrderShouldlyMessage(actual, expectedSortDirection, currentIndex, currentItem, customMessage).ToString());

--- a/src/Shouldly/ShouldlyExtensionMethods/ShouldBeGtLtTestExtensions.cs
+++ b/src/Shouldly/ShouldlyExtensionMethods/ShouldBeGtLtTestExtensions.cs
@@ -1,47 +1,46 @@
 ﻿using System;
 using System.Collections.Generic;
-using System.Diagnostics.CodeAnalysis;
 
 namespace Shouldly
 {
     public static partial class ShouldBeTestExtensions
     {
-        public static void ShouldBeGreaterThan<T>([AllowNull] this T actual, [AllowNull] T expected, IComparer<T> comparer, string? customMessage = null)
+        public static void ShouldBeGreaterThan<T>(this T? actual, T? expected, IComparer<T> comparer, string? customMessage = null)
         {
             actual!.AssertAwesomely(v => Is.GreaterThan(actual, expected, comparer), actual, expected, customMessage);
         }
 
-        public static void ShouldBeGreaterThan<T>([AllowNull] this T actual, [AllowNull] T expected, string? customMessage = null) where T : IComparable<T>?
+        public static void ShouldBeGreaterThan<T>(this T? actual, T? expected, string? customMessage = null) where T : IComparable<T>?
         {
             actual!.AssertAwesomely(v => Is.GreaterThan(v, expected), actual, expected, customMessage);
         }
 
-        public static void ShouldBeLessThan<T>([AllowNull] this T actual, [AllowNull] T expected, IComparer<T> comparer, string? customMessage = null)
+        public static void ShouldBeLessThan<T>(this T? actual, T? expected, IComparer<T> comparer, string? customMessage = null)
         {
             actual!.AssertAwesomely(v => Is.LessThan(actual, expected, comparer), actual, expected, customMessage);
         }
 
-        public static void ShouldBeLessThan<T>([AllowNull] this T actual, [AllowNull] T expected, string? customMessage = null) where T : IComparable<T>?
+        public static void ShouldBeLessThan<T>(this T? actual, T? expected, string? customMessage = null) where T : IComparable<T>?
         {
             actual!.AssertAwesomely(v => Is.LessThan(v, expected), actual, expected, customMessage);
         }
 
-        public static void ShouldBeGreaterThanOrEqualTo<T>([AllowNull] this T actual, [AllowNull] T expected, IComparer<T> comparer, string? customMessage = null)
+        public static void ShouldBeGreaterThanOrEqualTo<T>(this T? actual, T? expected, IComparer<T> comparer, string? customMessage = null)
         {
             actual!.AssertAwesomely(v => Is.GreaterThanOrEqualTo(actual, expected, comparer), actual, expected, customMessage);
         }
 
-        public static void ShouldBeGreaterThanOrEqualTo<T>([AllowNull] this T actual, [AllowNull] T expected, string? customMessage = null) where T : IComparable<T>?
+        public static void ShouldBeGreaterThanOrEqualTo<T>(this T? actual, T? expected, string? customMessage = null) where T : IComparable<T>?
         {
             actual!.AssertAwesomely(v => Is.GreaterThanOrEqualTo(v, expected), actual, expected, customMessage);
         }
 
-        public static void ShouldBeLessThanOrEqualTo<T>([AllowNull] this T actual, [AllowNull] T expected, IComparer<T> comparer, string? customMessage = null)
+        public static void ShouldBeLessThanOrEqualTo<T>(this T? actual, T? expected, IComparer<T> comparer, string? customMessage = null)
         {
             actual!.AssertAwesomely(v => Is.LessThanOrEqualTo(actual, expected, comparer), actual, expected, customMessage);
         }
 
-        public static void ShouldBeLessThanOrEqualTo<T>([AllowNull] this T actual, [AllowNull] T expected, string? customMessage = null) where T : IComparable<T>?
+        public static void ShouldBeLessThanOrEqualTo<T>(this T? actual, T? expected, string? customMessage = null) where T : IComparable<T>?
         {
             actual!.AssertAwesomely(v => Is.LessThanOrEqualTo(v, expected), actual, expected, customMessage);
         }

--- a/src/Shouldly/ShouldlyExtensionMethods/ShouldBeGtLtTestExtensions.cs
+++ b/src/Shouldly/ShouldlyExtensionMethods/ShouldBeGtLtTestExtensions.cs
@@ -7,42 +7,42 @@ namespace Shouldly
     {
         public static void ShouldBeGreaterThan<T>(this T? actual, T? expected, IComparer<T> comparer, string? customMessage = null)
         {
-            actual!.AssertAwesomely(v => Is.GreaterThan(actual, expected, comparer), actual, expected, customMessage);
+            actual.AssertAwesomely(v => Is.GreaterThan(actual, expected, comparer), actual, expected, customMessage);
         }
 
         public static void ShouldBeGreaterThan<T>(this T? actual, T? expected, string? customMessage = null) where T : IComparable<T>?
         {
-            actual!.AssertAwesomely(v => Is.GreaterThan(v, expected), actual, expected, customMessage);
+            actual.AssertAwesomely(v => Is.GreaterThan(v, expected), actual, expected, customMessage);
         }
 
         public static void ShouldBeLessThan<T>(this T? actual, T? expected, IComparer<T> comparer, string? customMessage = null)
         {
-            actual!.AssertAwesomely(v => Is.LessThan(actual, expected, comparer), actual, expected, customMessage);
+            actual.AssertAwesomely(v => Is.LessThan(actual, expected, comparer), actual, expected, customMessage);
         }
 
         public static void ShouldBeLessThan<T>(this T? actual, T? expected, string? customMessage = null) where T : IComparable<T>?
         {
-            actual!.AssertAwesomely(v => Is.LessThan(v, expected), actual, expected, customMessage);
+            actual.AssertAwesomely(v => Is.LessThan(v, expected), actual, expected, customMessage);
         }
 
         public static void ShouldBeGreaterThanOrEqualTo<T>(this T? actual, T? expected, IComparer<T> comparer, string? customMessage = null)
         {
-            actual!.AssertAwesomely(v => Is.GreaterThanOrEqualTo(actual, expected, comparer), actual, expected, customMessage);
+            actual.AssertAwesomely(v => Is.GreaterThanOrEqualTo(actual, expected, comparer), actual, expected, customMessage);
         }
 
         public static void ShouldBeGreaterThanOrEqualTo<T>(this T? actual, T? expected, string? customMessage = null) where T : IComparable<T>?
         {
-            actual!.AssertAwesomely(v => Is.GreaterThanOrEqualTo(v, expected), actual, expected, customMessage);
+            actual.AssertAwesomely(v => Is.GreaterThanOrEqualTo(v, expected), actual, expected, customMessage);
         }
 
         public static void ShouldBeLessThanOrEqualTo<T>(this T? actual, T? expected, IComparer<T> comparer, string? customMessage = null)
         {
-            actual!.AssertAwesomely(v => Is.LessThanOrEqualTo(actual, expected, comparer), actual, expected, customMessage);
+            actual.AssertAwesomely(v => Is.LessThanOrEqualTo(actual, expected, comparer), actual, expected, customMessage);
         }
 
         public static void ShouldBeLessThanOrEqualTo<T>(this T? actual, T? expected, string? customMessage = null) where T : IComparable<T>?
         {
-            actual!.AssertAwesomely(v => Is.LessThanOrEqualTo(v, expected), actual, expected, customMessage);
+            actual.AssertAwesomely(v => Is.LessThanOrEqualTo(v, expected), actual, expected, customMessage);
         }
     }
 }

--- a/src/Shouldly/ShouldlyExtensionMethods/ShouldBeRangeTestExtensions.cs
+++ b/src/Shouldly/ShouldlyExtensionMethods/ShouldBeRangeTestExtensions.cs
@@ -7,48 +7,48 @@ namespace Shouldly
 {
     public static partial class ShouldBeTestExtensions
     {
-        public static void ShouldBeOneOf<T>([AllowNull] this T actual, params T[] expected)
+        public static void ShouldBeOneOf<T>(this T? actual, params T[] expected)
         {
             ShouldBeOneOf(actual, expected, (string?)null);
         }
-        public static void ShouldBeOneOf<T>([AllowNull] this T actual, T[] expected, string? customMessage)
+        public static void ShouldBeOneOf<T>(this T? actual, T[] expected, string? customMessage)
         {
             // Enumerable.Contains on an array always tolerates null.
             if (!expected.Contains(actual!))
                 throw new ShouldAssertException(new ExpectedActualShouldlyMessage(expected, actual, customMessage).ToString());
         }
 
-        public static void ShouldBeOneOf<T>([AllowNull] this T actual, T[] expected, IEqualityComparer<T> comparer, string? customMessage = null)
+        public static void ShouldBeOneOf<T>(this T? actual, T[] expected, IEqualityComparer<T> comparer, string? customMessage = null)
         {
             // Enumerable.Contains on an array always tolerates null.
             if (!expected.Contains(actual!, comparer))
                 throw new ShouldAssertException(new ExpectedActualShouldlyMessage(expected, actual, customMessage).ToString());
         }
 
-        public static void ShouldNotBeOneOf<T>([AllowNull] this T actual, params T[] expected)
+        public static void ShouldNotBeOneOf<T>(this T? actual, params T[] expected)
         {
             ShouldNotBeOneOf(actual, expected, (string?) null);
         }
-        public static void ShouldNotBeOneOf<T>([AllowNull] this T actual, T[] expected, string? customMessage)
+        public static void ShouldNotBeOneOf<T>(this T? actual, T[] expected, string? customMessage)
         {
             // Enumerable.Contains on an array always tolerates null.
             if (expected.Contains(actual!))
                 throw new ShouldAssertException(new ExpectedActualShouldlyMessage(expected, actual, customMessage).ToString());
         }
 
-        public static void ShouldNotBeOneOf<T>([AllowNull] this T actual, T[] expected, IEqualityComparer<T> comparer, string? customMessage = null)
+        public static void ShouldNotBeOneOf<T>(this T? actual, T[] expected, IEqualityComparer<T> comparer, string? customMessage = null)
         {
             // Enumerable.Contains on an array always tolerates null.
             if (expected.Contains(actual!, comparer))
                 throw new ShouldAssertException(new ExpectedActualShouldlyMessage(expected, actual, customMessage).ToString());
         }
 
-        public static void ShouldBeInRange<T>([DisallowNull] this T actual, [AllowNull] T from, [AllowNull] T to, string? customMessage = null) where T : IComparable<T>
+        public static void ShouldBeInRange<T>([DisallowNull] this T actual, T? from, T? to, string? customMessage = null) where T : IComparable<T>
         {
             actual.AssertAwesomely(v => Is.InRange<T>(v, @from, to), actual, new { @from, to }, customMessage);
         }
 
-        public static void ShouldNotBeInRange<T>([DisallowNull] this T actual, [AllowNull] T from, [AllowNull] T to, string? customMessage = null) where T : IComparable<T>
+        public static void ShouldNotBeInRange<T>([DisallowNull] this T actual, T? from, T? to, string? customMessage = null) where T : IComparable<T>
         {
             actual.AssertAwesomely(v => !Is.InRange<T>(v, @from, to), actual, new { @from, to }, customMessage);
         }

--- a/src/Shouldly/ShouldlyExtensionMethods/ShouldBeTypeTestExtensions.cs
+++ b/src/Shouldly/ShouldlyExtensionMethods/ShouldBeTypeTestExtensions.cs
@@ -6,10 +6,10 @@ namespace Shouldly
 {
     public static partial class ShouldBeTestExtensions
     {
-        public static T ShouldBeAssignableTo<T>(this object? actual, string? customMessage = null)
+        public static T? ShouldBeAssignableTo<T>(this object? actual, string? customMessage = null)
         {
             ShouldBeAssignableTo(actual, typeof(T), customMessage);
-            return (T)actual!;
+            return (T)actual;
         }
 
         public static void ShouldBeAssignableTo(this object? actual, Type expected, string? customMessage = null)

